server/test/write_authzmodel.go

```diff
@@ -362,28 +491,78 @@ func TestWriteAuthorizationModel(t *testing.T, datastore storage.OpenFGADatastor
 					},
 				},
 			},
-			err: errors.RelationNotFound("owner", "", nil),
+			err: errors.CannotAllowMultipleReferencesToOneRelation,
 		},
 		{
-			name: "ExecuteWriteFailsIfUnknownRelationInTupleToUsersetComputedUserset",
+			name: "ExecuteWriteFailsIfIntersectionIncludesSameRelationTwice",
 			request: &openfgapb.WriteAuthorizationModelRequest{
-				TypeDefinitions: &openfgapb.TypeDefinitions{
-					TypeDefinitions: []*openfgapb.TypeDefinition{
-						{
-							Type: "repo",
-							Relations: map[string]*openfgapb.Userset{
-								"writer": {Userset: &openfgapb.Userset_This{}},
-								"viewer": {
-									Userset: &openfgapb.Userset_TupleToUserset{
-										TupleToUserset: &openfgapb.TupleToUserset{
-											Tupleset: &openfgapb.ObjectRelation{
-												Object:   "",
-												Relation: "writer",
-											},
+				StoreId: storeID,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "repo",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": {
+								Userset: &openfgapb.Userset_Intersection{
+									Intersection: &openfgapb.Usersets{Child: []*openfgapb.Userset{
+										{Userset: &openfgapb.Userset_ComputedUserset{
 											ComputedUserset: &openfgapb.ObjectRelation{
-												Object:   "$TUPLE_USERSET_OBJECT",
-												Relation: "owner",
+												Relation: "viewer",
 											},
+										}},
+										{Userset: &openfgapb.Userset_This{}},
+									}},
+								},
+							},
+						},
+					},
+				},
+			},
+			err: errors.CannotAllowMultipleReferencesToOneRelation,
+		},
+	}
+
+	ctx := context.Background()
+	logger := logger.NewNoopLogger()
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			cmd := commands.NewWriteAuthorizationModelCommand(datastore, logger)
+			resp, err := cmd.Execute(ctx, test.request)
+			require.ErrorIs(t, err, test.err)
+
+			if err == nil {
+				require.True(t, id.IsValid(resp.AuthorizationModelId))
+			}
+		})
+	}
+}
+
+func TestWriteAuthorizationModelWithRelationalTypes(t *testing.T, datastore storage.OpenFGADatastore) {
+	var tests = []struct {
+		name    string
+		request *openfgapb.WriteAuthorizationModelRequest
+		err     error
+	}{
+		{
+			name: "succeeds on a valid model with a type type",
+			request: &openfgapb.WriteAuthorizationModelRequest{
+				SchemaVersion: "1.1",
+				StoreId:       storeID,
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "user",
+					},
+					{
+						Type: "document",
+						Relations: map[string]*openfgapb.Userset{
+							"reader": {Userset: &openfgapb.Userset_This{}},
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"reader": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										{
+											Type: "user",
```

Comment:
I've been using help methods for things like this to shrink it from 3 lines to 1.

e.g. `relationref("user", "")`
or `relationref("group", "member")`

Reply:
Yeah, probably `typesystem`. I think that's a decent home, because it's referring to relations on a specific object type.

Reply:
Again, in a follow up PR. Is this `typesystem_testutils` stuff? How do we arrange all this?

Reply:
I have a follow up PR that will add this. I'll submit that once we merge this 👍